### PR TITLE
fix getMetadataFor detection

### PR DESCRIPTION
**Pull request by rande**
I am targetting this branch, because it is the stable one

## Changelog

```markdown
### Fixed
- Fix how metadata information are retrieved when admin information are dumped

```

## Subject

Fix how metadata information are retrieved when admin information are dumped

### CODE REVIEW

File: Command/ExplainAdminCommand.php [hunk 1]
```diff
@@ -89,9 +89,18 @@ public function execute(InputInterface $input, OutputInterface $output)
...
-            $metadata = $this->getContainer()->get('validator.validator_factory')->getMetadataFor($admin->getClass());
-        } else {
+            $factory = $this->getContainer()->get('validator.validator_factory');
+
+            if (method_exists($factory, 'getMetadataFor')) {
+                $metadata = $factory->getMetadataFor($admin->getClass());
+            }
+        }
+
+        // TODO: remove method check in next major release
```
Comment (greg0ire) [inline Command/ExplainAdminCommand.php]: Please replace `TODO` with `NEXT_MAJOR` (we will `grep` for that before releasing)

**Comment by OskarStark** [comment]
@greg0ire when i remember correctly, we have an issue or an open PR which tried to fix this, but which did not work as expected... 💭

**Comment by greg0ire** [comment]
Yes, this looks oddly familiar…

File: Command/ExplainAdminCommand.php [hunk 2]
```diff
@@ -89,9 +89,18 @@ public function execute(InputInterface $input, OutputInterface $output)
...
-            $metadata = $this->getContainer()->get('validator.validator_factory')->getMetadataFor($admin->getClass());
-        } else {
+            $factory = $this->getContainer()->get('validator.validator_factory');
+
+            if (method_exists($factory, 'getMetadataFor')) {
+                $metadata = $factory->getMetadataFor($admin->getClass());
+            }
+        }
+
+        // NEXT_MAJOR: remove method check in next major release
```
Comment (OskarStark) [inline Command/ExplainAdminCommand.php]: if we do this we should trigger a deprecation here

Comment (rande) [inline Command/ExplainAdminCommand.php]: No; it is not part of the user land; it is just a command to dump information.

**Comment by greg0ire** [comment]
Here is the history : 

```diff
commit 3e7ad359411a5f4ae119d9dc3ef26b3e44716fd1
Author: Thomas Rabaix <thomas.rabaix@gmail.com>
Date:   Mon Nov 23 19:00:57 2015 +0100

    Add support for SF3.0 (WIP)

diff --git a/Command/ExplainAdminCommand.php b/Command/ExplainAdminCommand.php
index 11cfca1..e35235e 100644
--- a/Command/ExplainAdminCommand.php
+++ b/Command/ExplainAdminCommand.php
@@ -89,16 +89,14 @@ public function execute(InputInterface $input, OutputInterface $output)
             $output->writeln(sprintf('  - % -25s  % -15s % -15s', $name, $fieldDescription->getType(), $fieldDescription->getTemplate()));
         }
 
-        $validator = $this->getContainer()->get('validator');
-        // TODO: Remove conditional method when bumping requirements to SF 2.5+
-        if (method_exists($validator, 'getMetadataFor')) {
-            $metadata = $validator->getMetadataFor($admin->getClass());
+        if ($this->getContainer()->has('validator.validator_factory')) {
+            $metadata = $this->getContainer()->get('validator.validator_factory')->getMetadataFor($admin->getClass());
         } else {
-            $metadata = $validator->getMetadataFactory()->getMetadataFor($admin->getClass());
+            $metadata = $this->getContainer()->get('validator')->getMetadataFor($admin->getClass());
         }
 
         $output->writeln('');
-        $output->writeln('<comment>Validation Framework</comment> - http://symfony.com/doc/2.0/book/validation.html');
+        $output->writeln('<comment>Validation Framework</comment> - http://symfony.com/doc/3.0/book/validation.html');
         $output->writeln('<info>Properties constraints</info>');
 
         if (count($metadata->properties) == 0) {

commit baf9d425f67486bc1ca6612fa15adf703a8d1e64
Author: Sullivan SENECHAL <soullivaneuh@gmail.com>
Date:   Fri May 1 12:21:44 2015 +0200

    Fix deprecated getMetadataFactory method usage

diff --git a/Command/ExplainAdminCommand.php b/Command/ExplainAdminCommand.php
index 08800f8..19bd482 100644
--- a/Command/ExplainAdminCommand.php
+++ b/Command/ExplainAdminCommand.php
@@ -90,8 +90,13 @@ public function execute(InputInterface $input, OutputInterface $output)
             $output->writeln(sprintf('  - % -25s  % -15s % -15s', $name, $fieldDescription->getType(), $fieldDescription->getTemplate()));
         }
 
-        $validatorFactory = $this->getContainer()->get('validator')->getMetadataFactory();
-        $metadata = $validatorFactory->getMetadataFor($admin->getClass());
+        $validator = $this->getContainer()->get('validator');
+        // TODO: Remove conditional method when bumping requirements to SF 2.5+
+        if (method_exists($validator, 'getMetadataFor')) {
+            $metadata = $validator->getMetadataFor($admin->getClass());
+        } else {
+            $metadata = $validator->getMetadataFactory()->getMetadataFor($admin->getClass());
+        }
 
         $output->writeln('');
         $output->writeln('<comment>Validation Framework</comment> - http://symfony.com/doc/2.0/book/validation.html');

commit 1394100cd8d19aacbde1c70a2e2e529523920f57
Author: Thomas Rabaix <thomas.rabaix@sonata-project.org>
Date:   Fri Jun 7 22:52:49 2013 +0200

    Fix deprecated call

diff --git a/Command/ExplainAdminCommand.php b/Command/ExplainAdminCommand.php
index 6d310b4..c2bbac4 100644
--- a/Command/ExplainAdminCommand.php
+++ b/Command/ExplainAdminCommand.php
@@ -87,7 +87,7 @@ public function execute(InputInterface $input, OutputInterface $output)
         }
 
         $validatorFactory = $this->getContainer()->get('validator.mapping.class_metadata_factory');
-        $metadata = $validatorFactory->getClassMetadata($admin->getClass());
+        $metadata = $validatorFactory->getMetadataFor($admin->getClass());
 
         $output->writeln('');
         $output->writeln('<comment>Validation Framework</comment> - http://symfony.com/doc/2.0/book/validation.html');
```

**Comment by greg0ire** [comment]
Related : #4045

**Comment by rande** [comment]
All green can we merge ?

**Comment by greg0ire** [comment]
LGTM

**Comment by rande** [comment]
@OskarStark @Soullivaneuh can we merge this PR ?

**Comment by greg0ire** [comment]
@sonata-project/contributors , final review and merge?

**Comment by core23** [comment]
LGTM

**Comment by greg0ire** [comment]
Thanks @rande

**Pull request merged**